common/kotlinx-coroutines-core-common/src/JobSupport.kt

```diff
@@ -1077,7 +1078,9 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
             val state = this.state
             if (state !is Incomplete) {
                 // already complete -- just return result
-                if (state is CompletedExceptionally) throw state.cause
+                if (state is CompletedExceptionally) { // Slow path to recover stacktrace
+                    recoverAndThrow(state.cause)
```

Comment:
Such check should be inserted in every fast-path method of the channel

Reply:
Looks ok to me. It does not affect that speed of fast-path (still no state-machine) and only slows down exceptions.